File: day_02/day2.py
from aoc import aoc

moves = {
    "A": "rock",
    "B": "paper",
    "C": "scissors",
    "X": "rock",
    "Y": "paper",
    "Z": "scissors"
}
scores = {
    "rock": 1,
    "paper": 2,
    "scissors": 3
}
# 6 = win, 3 = draw, 0 = lose
# x=me, y=them
outcome = [
    [3, 6, 0],
    [0, 3, 6],
    [6, 0, 3]]


def get_score(them, me):
    my_move = moves[me]
    their_move = moves[them]
    return scores[my_move] + outcome[scores[their_move] - 1][scores[my_move] - 1]


def add_scores(data):
    goes = [row.split(" ") for row in data]
    return sum([get_score(*row) for row in goes])


map_outcome_to_score = {
    "X": 0,
    "Y": 3,
    "Z": 6
}


def add_scores2(data):
    goes = [row.split(" ") for row in data]
    return sum([get_score(g[0], get_move(*g)) for g in goes])


def get_move(their_move, winlosedraw):
    scoreneeded = map_outcome_to_score[winlosedraw]
    them = moves[their_move]  # name of their move
    rowno = scores[them] - 1  # row in matrix
    moveneeded = outcome[rowno].index(scoreneeded)
    return "ABC"[moveneeded]


def run(file):
    print("Part 1")
    data = aoc.loadlines(file)
    print(add_scores(data))
    print("Part 2")
    print(add_scores2(data))


if __name__ == "__main__":
    print("Test Data")
    run("test.txt")
    print("Actual Data")
    run("data.txt")

File: day_02/day2.jl
#=
day2:
- Julia version: 
- Author: mike.downey
- Date: 2022-12-04

This version is very similar to the original python code.
=#

# file = open("day_02/test.txt")
file = open("day_02/data.txt")
data = readlines(file)

# Part 1
println("Part 1")

outcome = [
    [3 6 0],
    [0 3 6],
    [6 0 3]]

moves = Dict(
    "A" => "rock",
    "B" => "paper",
    "C" => "scissors",
    "X" => "rock",
    "Y" => "paper",
    "Z" => "scissors"
)

scores = Dict(
    "rock" => 1,
    "paper" => 2,
    "scissors" => 3
)

function get_score(them, me)
    my_move = moves[me]
    their_move = moves[them]
    return scores[my_move] + outcome[scores[their_move]][scores[my_move]]
end

function split_lines(lines, separator)
    output = []
    for line in lines
        println(line)
        row = split(line, separator)
        push!(output, row)
    end
    return output
end

function add_scores(lines)
    score = 0
    for row in lines
        if length(row)>0
            l = split(row, " ")
            score += get_score(l[1], l[2])
        end
    end
    return score
end

println(add_scores(data))

# Part 2
println("Part 2")

map_outcome_to_score = Dict(
    "X"=> 0,
    "Y"=> 3,
    "Z"=> 6
)

function get_move(their_move, winlosedraw)
    scoreneeded = map_outcome_to_score[winlosedraw]
    them = moves[their_move]  # name of their move
    rowno = scores[them]  # row in matrix
    moveneeded = findfirst(isequal(scoreneeded), outcome[rowno])
    # Returns a 'CartesianIndex' which we can use to extract from an array but not a string.
    return ["A" "B" "C"][moveneeded]
end

score = 0
for row in data
    global score
    r = split(row, " ")
    move = get_move(r[1], r[2])
    score += get_score(r[1], move)
end

println(score)